### DISCUSSION TITLE
Control the contact between a rigid ball and a flexible object

### DISCUSSION THREAD

**Discussion by zhanglepy** [sofa-framework/sofa#5022] [SofaPython3] [answered]
hello everyone,
I recently used SofaPython3 to control the contact between a rigid ball and a flexible object. By controlling the position of the rigid body, I slowly approached and squeezed the flexible object, but I found that after contact, the position control of the rigid body failed, and I could not continue to control the rigid body to further squeeze the flexible object. Why doesn't the program for controlling rigid body motion work? Below are my complete program and files. Could you please help me solve this problem. Looking forward to anyone's reply!
![01](https://github.com/user-attachments/assets/9306d714-d0fd-4ddd-99e5-598f8d8bab7f)
![02](https://github.com/user-attachments/assets/e9e63b37-189e-47c6-a39f-84904772e6ce)
[test.zip](https://github.com/user-attachments/files/17130060/test.zip)

**Comment by zhanglepy** [comment]
0.49
0.48
0.47
0.45999999999999996
0.44999999999999996
0.43999999999999995
0.42999999999999994
0.41999999999999993
0.4099999999999999
0.3999999999999999
0.3899999999999999
0.3799999999999999
0.3699999999999999
0.3599999999999999
0.34999999999999987
0.33999999999999986
0.32999999999999985
0.31999999999999984
0.30999999999999983
0.2999999999999998
0.2899999999999998
0.2799999999999998
0.2727423048200836
0.27221801798203443
0.27217616632185054
0.27217280025655644
0.2721725293684527
0.2721725075698832
0.2721725147368163
0.27217255086718745
0.2721726159452344
0.2721727099528323
0.27217283286979366
0.2721729846738745

This is the position information of the output motion direction. You will find that the height has been decreasing recently, but after the rigid body contacts the flexible object, the height does not decrease anymore.

**Comment by zhanglepy** [comment]
Below are my complete program and files. 
[test.zip](https://github.com/user-attachments/files/17130104/test.zip)

**Comment by zhanglepy** [comment]
Looking forward to anyone's reply!

**Comment by bakpaul** [comment, marked as answer]
Hello ! 

The problem here is that you are mistaking imposing position with modification of its position data field. This data-field is overridden by the solver. When you move it yourself, if it is interacting with nothing, the solver will not modify it because the increase of velocity of this time-step is null, because no force is applied on it. But when you enter in contact with the object, nothing is holding it at the place you are placing it, so when solving for the contact, it is way easier to push on the sphere than on the model, so its position seems kind of overridden. 

What you need here, is to impose the position of the sphere in a way, telling the solver that its position is strongly constrained. For this you have two possibilities : 
- Use another Lagrangian based constraint (those are the same as the one used for the contact here). This is done by adding a `BilateralLagrangianConstraint`. You need to add another node to your scene, but a node that is not simulated (with no ode solver and only a `MechanicalObject`). This is the `MechanicalObject` where you'll put the position you want to impose, then you add a `BilateralLagrangianConstraint` between this `MechanicalObject` and the sphere one. This way, you add an equality constraintbetween your sphere position and the one in the `MechanicalObject`, and it is a strong one because se added `MechanicalObject` is not simulated and thus cannot move. 
- Use a projective constraint on your sphere and remove its ConstraintCorrection. Removing the ConstraintCorrection will not allow the constraint solver to move it, so its position will be fixed during the solving. Then you can add a `LinearMovementProjectiveConstraint` where you can define a path that your object takes. But it might also work only with your controller. 

Kno that imposing a position like that might lead to instabilities, compressing an object is not very easy in term of solving. Moreover I see that you are fixing the liver with `FixedConstraint`: when you'll be pushing to much on the liver you'll see vibrations of your models, this is due to the nature of the `FixedConstraint` that you use. If you are using the v24.06 or the master version, I strongly suggest you to use the `FixedLagrangianConstraint` instead, this will use real Lagrangian constraints instead. It might run slowly but you'll have no vibration. But again, imposing unsolvable constraints will lead to numerical issues, makes sure you are not over constraining your model, otherwise the solver will never converge and you'll get a wrong behavior. 

Hope this was helpful

**Reply by zhanglepy** [reply]
Thank you for your reply. I will make the modifications according to your suggestions.
